tidy3d/components/eme/simulation.py

```diff
@@ -542,9 +553,34 @@ def _post_init_validators(self) -> None:
         self._validate_modes_size()
         self._validate_sweep_spec()
         self._validate_symmetry()
+        self._validate_constraint()
         # self._warn_monitor_interval()
         log.end_capture(self)
 
+    def _validate_constraint(self):
+        """Constraint can be slow with too many modes. Warn in this case."""
+        constraint = self.constraint
+        max_num_modes = self.max_num_modes
+        if constraint is not None and max_num_modes > WARN_CONSTRAINT_NUM_MODES:
+            log.warning(
+                f"The simulation has 'constraint='{constraint}'', and the maximum "
+                f"number of EME modes in the simulation is '{max_num_modes}'. "
+                f"Using more than '{WARN_CONSTRAINT_NUM_MODES}' modes together with "
+                "a constraint can significantly slow down the simulation. Consider "
+                "reducing the number of modes or setting 'constraint=None'."
+            )
+
+    def _validate_port_offsets(self):
+        """Port offsets cannot jointly exceed simulation length."""
```

Comment:
does this need to take `self.center` into account?

Maybe I dont quite understand, is this kind because the ports are guaranteed to be outside of the simulation if their offset is greater than the simulation size?

Reply:
The `port_offsets` defines the amount by which the ports (the beginning and end of the EME grid) are offset inward from the simulation boundaries. If the sum of the offsets is larger than the simulation size, the EME grid will have negative extent.